New UI KitchenSink Component Speedo / Gauge

qdittrich opened the pull request:
### New UI kitchenSink Component: Gauge
https://www.ilias.de/docu/goto_docu_wiki_wpage_4929_1357.html#ilPageTocA212
Old PR: #786 (Speedo)

Todo's:
- [x] Extend the tests..
- [x] Test the Less/CSS on various browsers.
- [x] Rename Speedo Familiy.
- [x] Change "Goal" to "Required".
- [x] Change "Score" to "Main Value".
- [x] Change "Diagnostic Score" to "Comparision Value".
- [x] Make "Responsive" as standard variant of Component.
- [x] Make "Fixed Size" as secondary variant of Component.
- [x] Check and Modify colors, text sizes and element sizes.
- [x] Implement more phpunit tests.
- [x] Modify descriptions, purposes, rules, etc.
- [x] Add example screenshots to PR.
- [x] Find final name
- [x] Make background more "horseshoe-like".
- [x] Optimize marker position and bar stretching.
- [x] Finalise the visualisation.
- [x] Check function descriptions.
- [x] Check documentation texts.
- [x] Check browser compatibility again.

As suggested in the Jour Fixe (04 DEC 2017), I reworked some parts and changed the family name of the speedo component.

The new family name "Gauge" is not final yet. I changed the name to make the new implementation step more visible and "Gauge" sounded good to me, also with regard to possible future Sub-Components. The discussion about the final name is still open because the majority of JF participants could not agree on one specific name.

qdittrich committed:
Added UI kitchenSink Speedo Interfaces

qdittrich committed:
Added UI kitchenSink Speedo Implementation

qdittrich committed:
Added UI kitchenSink Speedo examples, styling and test

qdittrich committed:
Applied requested changes

qdittrich committed:
Reworked KS Component Speedo / Gauge

qdittrich commented:
**Description Gauge**
![ks_chart_gauge](https://user-images.githubusercontent.com/33311456/33611857-1658ef86-d9d0-11e7-999a-fcd89f331d59.png)
**Description Gauge Standard**
![ks_chart_gauge_standard](https://user-images.githubusercontent.com/33311456/33611873-21a90024-d9d0-11e7-9493-0de2b473389a.png)
**Description Gauge Fixed Size**
![ks_chart_gauge_fixed](https://user-images.githubusercontent.com/33311456/33611879-2a26f5d0-d9d0-11e7-901c-8c7cd1ad8872.png)
**Description Gauge Mini**
![ks_chart_gauge_mini](https://user-images.githubusercontent.com/33311456/33611904-391d5ebc-d9d0-11e7-98cc-e370b77087e5.png)

**Example Standard / Fixed Size: Base**
![ks_chart_gauge_standard_base](https://user-images.githubusercontent.com/33311456/33611916-45d0371a-d9d0-11e7-8f28-5add67179d0c.png)
**Example Standard / Fixed Size: Max Configuration**
![ks_chart_gauge_standard_max_conf](https://user-images.githubusercontent.com/33311456/33611927-5128a6c4-d9d0-11e7-8de6-2218baff3f94.png)
**Example Standard / Fixed Size: Reached Required**
![ks_chart_gauge_standard_reached_required](https://user-images.githubusercontent.com/33311456/33612063-c1d7281e-d9d0-11e7-9af9-00cad2ed555f.png)


**Example Mini: Base**
![ks_chart_gauge_mini_base](https://user-images.githubusercontent.com/33311456/33611939-5d2d3598-d9d0-11e7-8490-5e68d4737f52.png)
**Example Mini: Max Configuration**
![ks_chart_gauge_mini_max_conf](https://user-images.githubusercontent.com/33311456/33611956-69632ee4-d9d0-11e7-9dd6-4be85f39f155.png)
**Example Mini: 100% Reached**
![ks_chart_gauge_mini_reached](https://user-images.githubusercontent.com/33311456/33612098-d946dd28-d9d0-11e7-9388-c781fb3a01ab.png)

qdittrich committed:
Modified styling and added marker to mini version

qdittrich commented:
**Short overview of new styling:**
![overview](https://user-images.githubusercontent.com/33311456/33874587-99cab684-df1f-11e7-9f17-9d814011e4b9.png)

qdittrich commented:
There are still two open points for the Discussion:
1. Naming: Will it be Gauge or do we want to change that name?
2. Visualization: Is everyone satisfied with the current visualization?

Because the deadline is on friday, the 15th, we should agree on this points soon. So, if anyone wants to participate in the discussion or want to give a suggestion, it would be great to get the feedback until afternoon at thurday, the 14th.

atoedt commented on src/UI/Component/Chart/Factory.php:
```diff
@@ -28,4 +28,32 @@
+     * ---
+     * description:
+     *   purpose: >
+     *     Gauges are used to display a progress or performance.
```
Please consider to name the family article "Progress indicator" to make it more accommodating and open to other entries. 
**Learning Progress Indicator** 
- Standard Learning Progress Indicator
- Fixed Size  Learning Progress Indicator
- Mini  Learning Progress Indicator

matthiaskunkel replied:
We should not choose a name for such an element that is already used by a specific feature in ILIAS. Otherwise we could also call it Learning Objective Indicator - because this is where it will appear first. In my understanding, we already have learning progress indicators - our so-called "Pöppel" in grey, yellow, green and red. These are learning progress indicators. 
Chris' suggestion to use something with "...meter" convinces me - because this is how it looks like. And it has no relation to an existing feature - same like all the other UI components we already have.

atoedt replied:
The Pöppel is a Learning Progress Status Display. The Pöppel is not measuring progress but indicates a status. 
However the Pöppel has not KS-Entry

matthiaskunkel replied:
If something indicates something, it is an indicator. Measuring is done by none of our UI elements but of the programme itself. The general difference between the pöppel and the meter is that the first can only display one status (presented by colour) while the second allows to present a value on a scale. This is why such an element is good for all cases where percentages or points should be shown and where an additional status related on the reached percentage/points should be presented (bar turns to green if required level is reached for example). Score Meter or Progress Meter would be fine for me, too.

atoedt replied:
The Progress Bar for the download is also a Progress Meter with a max value and a current value. We have currently just this element. There are several obvious candidates: To allow distinguishing between the Speedo, the Pöppel, and the Download Progress bar the names must not be too generic. There will be more Progress Indicators at some point in time.

matthiaskunkel commented:
New styling is fine. I am satisfied with the current visualisation. 

Concerning the element's name: I have asked our English translator (native) and here is his feedback: 
"gauge is an ugly word that is going to lead to mispronunciations among foreign learners... 'I have have tried the ga-uge and....(everybody falls around laughing). I suggest some form of 'meter'. Speedo is short for speedometer. Performance Meter? Learnometer?"
and
"'your score' and 'required score' are fine. As computer games often measure scores in percentages - it is not at all odd."
So my favourite would be "learnometer" because I do not like the connotation of 'performance' in relation to humans (term should be restricted to machines and economics). Or we use "Speedometer" in respect of the visual presentation.

atoedt commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     * ---
+     * description:
+     *   purpose: >
+     *     The standard gauge is usually the tool of choice. It is to be used, if the speedo needs to adapt
```
The Standard Learning Progress Indicator informs users about their progress compared to a the required maximum

atoedt commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     * description:
+     *   purpose: >
+     *     The standard gauge is usually the tool of choice. It is to be used, if the speedo needs to adapt
+     *     it's size depending on the screen size due to a more limited amount of space.
```
Leave out references to responisiveness in the purpose statement

atoedt commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     *
+     * rules:
+     *   composition:
+     *     1: Standard Gauges MAY contain a comparision value. It MUST be a numeric value between 0 and the maximum. It is represented as the second bar.
```
Ambivalence "It"
Replace by: If there is a comparison value it MUST...

qdittrich replied:
You are right, this will be much better to read.

atoedt commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     * ---
+     * description:
+     *   purpose: >
+     *     The fixedSize gauge is mostly identical with the standard. It is used if
```
replace by "The Fixed Size Learning Progress Indicator ensures that the element is rendered exactly as set regardless of the screen size."

qdittrich replied:
Very good phrasing! This sounds good to me :-)

atoedt commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     * ---
+     * description:
+     *   purpose: >
+     *     The mini gauge is used, if it needs to be as small as possible, like in an heading. It
```
The Mini Learning Progress Indicator is used to impart a reduced amount of information. The Mini Learning Progress Indicator is accompanies other structured data (i.e. heading or list item) and prevents informational overload from users.

qdittrich replied:
Im not sure about this formulation. The only information that is lost, is the output of text. The point of this element is that it will always be like a "miniature" of the component. Therefore I don't understand it as a reduction in functionality or information  but as a shrinked version.

atoedt replied:
Then we do not need a Mini but just the Fixed Size set to very small.

atoedt commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     * description:
+     *   purpose: >
+     *     The mini gauge is used, if it needs to be as small as possible, like in an heading. It
+     *     is used to display only a single progress or performance indicator.
```
This is NOT about size, if it was about size I could take the Fixed Size one and make it small. Which is not what I do. This is about the reduction of information nested in an already information-rich locus.

qdittrich replied:
See above comment

qdittrich commented:
Thank you for your feedback.

Because of the name, we should not forget the objections from the last JF.
In addition, at the JF, was mentioned that not everyone could recognize a speedometer in the visualization of the component.

Maybe something like ScoreMeter, ProgressChart or simply ProgressIndicator?

The component can not only indicate learning progress. It can also be used for any other values, that have a relation to each other. Therefore, the reference to the word "learning" seems not be optimal to me.

atoedt commented:
Please do not use CamelCases in names.

qdittrich commented:
Why should we not use CamelCase in the name? Some KS Components already do this.

Amstutz commented:
> Why should we not use CamelCase in the name? Some KS Components already do this.

On a technical level (interface etc.) CamelCase will be the way to go. In the documentation and UI, we transform CamelCase to Camel Case. See existing items. I think this is just a case of two different perspectives on the same thing.

qdittrich commented:
@Amstutz Ahh, i see. I thought @atoedt was talking about the "technical names". Sorry for the missunderstanding.

atoedt commented:
If text is displayed to a  user and the language is English, we committed ourselves to using proper BE. 
No KS element I am aware of uses CamelCases.

klees commented on src/UI/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return int|float
+     */
+    public function getComparision();
```
getComparision  -> getComparison

klees commented on src/UI/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,59 @@
+interface Standard extends Gauge {
+
+    /**
+     * Get comparision value
```
comparision -> comparison

qdittrich replied:
So many typos on the same word... I will correct this.

klees commented on src/UI/Component/Chart/Gauge/Gauge.php:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @return bool
+     */
+    public function hasComparisonValue();
```
I still don't think this method makes sense in that way here. It's super nitpicky, tough.

I mean: look at this interface as found here. By the means provided by it, I could only tell, that some object implementing that interface *has* a comparison value but I cannot know *what that value is*, because the interface provides no method to get that value.

On the other hand, the interface `Standard` defines a `getComparison` method, but does not define what happens when I call it in the case that `hasComparisonValue() == false`. I also cannot know that I can call this method on some `Gauge`-object, even if `hasComparisonValue() == true`. I would need an explicit check on `instanceof Standard` on the object to know the method really exists.

So in fact `hasComparisonValue` provides only very little value to users of the `Gauge`-interface. The really useful check is `instanceof Standard`. On the other hand I could build implementations of `Standard` that show contradictory behaviour regarding `hasComparisonValue` and `getComparisonValue`.

qdittrich replied:
I see what you mean. I will look again at this function. Maybe I will throw it away and just test in the renderer if the comparision value is not null.

klees commented on src/UI/Implementation/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * @inheritdoc
+     */
+    public function getComparision()
```
Comparision again.

klees reviewed (changes requested):
Really Nice work, thx. The pics in the PR are very useful. I reiterated one argument from the last PR and found a (repeated) typo.

klees commented on src/UI/Component/Chart/Gauge/Factory.php:
```diff
@@ -0,0 +1,83 @@
+     * @param int|float|null $comparision Comparision value to be displayed by second bar.
+     * @return \ILIAS\UI\Component\Chart\Gauge\Standard
+     */
+    public function standard($maximum, $main, $required = null, $comparision = null);
```
comparision again

klees commented on src/UI/Component/Chart/Gauge/Gauge.php:
```diff
@@ -0,0 +1,52 @@
+     * This value represents the required amount that is needed, to fulfill the objective.
+     * If this value is not set, it defaults to the maximum.
+     *
+     * @return int|float
```
This might be `null` as well.

klees commented on src/UI/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Get main text value
+     *
+     * @return string
```
This could be `null` as well, or am I missing something?

qdittrich replied:
I had to check this again and you are right.

klees commented on src/UI/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,59 @@
+     * @param string $text
+     * @return \ILIAS\UI\Component\Chart\Gauge\Gauge
+     */
+    public function withRequiredText($text);
```
I think this also might be `null`.

klees commented on src/UI/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * Get required text value
+     *
+     * @return string
```
I think this also might be `null`.

qdittrich replied:
yes, like above it may be null.

klees commented on src/UI/Component/Chart/Gauge/Standard.php:
```diff
@@ -0,0 +1,59 @@
+     *
+     * This value is represented as the second gauge bar.
+     *
+     * @return int|float
```
There definitely are cases there is no such value.

matthiaskunkel commented:
Jour Fixe, 18 DEC 2017: We highly appreciate this PR and accept the interface. The UI name should be "Progress Meter". PR can be merged after implementation has been reviewed by Timon and Richard.

qdittrich committed:
Renamed KS Component Speedo

New Name: ProgressMeter
Also fixed typo "comparision" => "comparison"

qdittrich committed:
Merge remote-tracking branch 'origin/trunk' into speedo

qdittrich committed:
Modified descriptions and documentation texts

Also removed function "hasComparisonValue()"

qdittrich committed:
Merge remote-tracking branch 'origin/trunk' into speedo

qdittrich committed:
updated data.json and delos.css

qdittrich committed:
Fixed KS Porgress Meter example filenames

qdittrich commented:
I changed the name. Also I modified the documentation and function descriptions.
The typos are fixed and I checked it twice, to see if there are some more. To me, it seems to be typo clean now.

I merged the current trunk, to get rid of the merge conflicts. The only two conflicts were the delos.css and data.json files.

Below are two more screenshots, so you can see the current state of the component documentation.
![image](https://user-images.githubusercontent.com/33311456/34198105-7ab68944-e569-11e7-933c-c316ca2b120a.png)

![image](https://user-images.githubusercontent.com/33311456/34198262-016eb452-e56a-11e7-8619-9e83214fff7d.png)

Amstutz commented:
While going through some old PR's assigned to me I realised with great shock that this has not been merged yet. @klees is still marked as requesting changes. However as far as I can tell, the requested changes from the last review are all worked in. The less looks and the examples look fine. The tests all pass. I will merge this to complete this. 

If there are further things to be discussed, please open a new PR and request new suggestions there.

Finally: @qdittrich again, thanks again for this great work!
